packages/hemi-tunnel-actions/src/utils.ts

```diff
@@ -0,0 +1,110 @@
+import { EventEmitter } from 'events'
+import {
+  Address,
+  Chain,
+  ChainContract,
+  Hash,
+  isAddress,
+  PublicClient,
+} from 'viem'
+
+import { DepositEvents } from './types'
+
+type DefaultEventMap = [never]
+type EventMap<T> = Record<keyof T, unknown[]> | DefaultEventMap
+
+/**
+ * Converts a function that accepts an event emitter and returns a promise into
+ * a "promise event" object, which exposes the promise and the event emitter, both
+ * in a sync fashion.
+ */
+export const toPromiseEvent = function <
+  T extends EventMap<T> = DefaultEventMap,
+>(callback: (emitter: EventEmitter<T>) => Promise<void>) {
+  const emitter = new EventEmitter<T>()
+
+  // eslint-disable-next-line promise/no-callback-in-promise
+  const promise = Promise.resolve().then(() => callback(emitter))
+
+  return { emitter, promise }
+}
+
+export const getL1StandardBridgeAddress = function ({
+  l1Chain,
+  l2Chain,
+}: {
+  l1Chain: Chain
+  l2Chain: Chain
+}) {
+  const l1StandardBridge = (
+    l2Chain.contracts?.l1StandardBridge as {
+      [sourceId: number]: ChainContract
+    }
+  )?.[l1Chain.id].address as Address | undefined
+  if (!l1StandardBridge) {
+    throw new Error(
+      `Chain ${l2Chain.id} is missing L1StandardBridge contract for source chain ${l1Chain.id}`,
+    )
+  }
+  return l1StandardBridge
+}
+
+export const handleWaitDeposit = async function <T extends DepositEvents>({
+  emitter,
+  hash,
+  publicClient,
+}: {
+  emitter: EventEmitter<EventMap<T>>
+  hash: Hash | undefined | void
+  publicClient: PublicClient
+}) {
+  if (!hash) {
+    return
+  }
```

Comment:
This check should be hoisted to the caller as it is the way to handle the call to `writeContract()` failed. Same as done here in L74-76. It "logically" belongs with the previos call.

Reply:
Done [03d3a22](https://github.com/hemilabs/ui-monorepo/pull/1119/commits/03d3a2230215d9db9837f7c15efbc6bd5397ac2d)